ocsp/responder/redis/checked_redis_source.go

```diff
@@ -0,0 +1,139 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/go-gorp/gorp/v3"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/db"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/ocsp/responder"
+	"github.com/letsencrypt/boulder/revocation"
+	"github.com/letsencrypt/boulder/sa"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+// dbSelector is a limited subset of the db.WrappedMap interface to allow for
+// easier mocking of mysql operations in tests.
+type dbSelector interface {
+	SelectOne(holder interface{}, query string, args ...interface{}) error
+	WithContext(ctx context.Context) gorp.SqlExecutor
+}
+
+// rocspSourceInterface expands on responder.Source by adding a private signAndSave method.
+// This allows checkedRedisSource to trigger a live signing if the DB disagrees with Redis.
+type rocspSourceInterface interface {
+	Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error)
+	signAndSave(ctx context.Context, req *ocsp.Request, cause string) (*responder.Response, error)
+}
+
+type checkedRedisSource struct {
+	base    rocspSourceInterface
+	dbMap   dbSelector
+	counter *prometheus.CounterVec
+	log     blog.Logger
+}
+
+func NewCheckedRedisSource(base *redisSource, dbMap dbSelector, stats prometheus.Registerer, log blog.Logger) *checkedRedisSource {
```

Comment:
Please add a comment that denotes the significance of `nil` being returned here. It might make more sense to change the signature of this function to return an error though.